File: python/ch05_agent/agent_ch05/core/hooks.py
"""第四章 Hook 生命周期。

Java 对照：`HookRegistry` 类似一个按事件分组的观察者注册表，
`HookContext` 和 `HookResult` 类似经过校验的不可变 DTO。回调只能声明影响，
不能直接修改 Agent 循环，所以扩展逻辑不会重新变成一堆 if/else。
"""

import inspect
from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass
from typing import Literal

from .messages import (
    AssistantMessage,
    ChatMessage,
    SystemMessage,
    ToolMessage,
    UserMessage,
    system_message,
    user_message,
)
from .permissions import PERMISSION_BEHAVIORS, PermissionBehavior
from .tools import PreparedToolCall, ToolResult, copy_prepared_tool_call, copy_tool_result

HookEvent = Literal["UserPromptSubmit", "PreToolUse", "PostToolUse", "Stop"]
HOOK_EVENTS: tuple[HookEvent, ...] = ("UserPromptSubmit", "PreToolUse", "PostToolUse", "Stop")


class HookContractError(Exception):
    """Hook 输入、输出或事件字段违反契约时抛出的领域异常。"""


def _is_event(value: object) -> bool:
    return value in HOOK_EVENTS


def _is_prepared(value: object) -> bool:
    return isinstance(value, PreparedToolCall) and value.error is None and value.definition is not None and value.arguments is not None


@dataclass(frozen=True, slots=True)
class HookContext:
    """某次回调能看到的最小事件上下文。

    Python 的 `None` 类似 Java 的 `null`，但这里不是任意字段都能为 None：
    `__post_init__` 会根据事件强制检查字段归属，防止 Hook 读取错误阶段的数据。
    """

    event: HookEvent
    message: ChatMessage | None = None
    prepared: PreparedToolCall | None = None
    result: ToolResult | None = None
    history: tuple[ChatMessage, ...] = ()
    stop_hook_active: bool = False

    def __post_init__(self) -> None:
        if not _is_event(self.event):
            raise HookContractError("event 必须是受支持的 HookEvent")
        if not isinstance(self.history, tuple) or not all(isinstance(item, (SystemMessage, AssistantMessage, UserMessage, ToolMessage)) for item in self.history):
            raise HookContractError(f"{self.event} history 必须全部是合法消息")
        if not isinstance(self.stop_hook_active, bool):
            raise HookContractError("stop_hook_active 必须是 bool")
        if self.event == "UserPromptSubmit":
            if not isinstance(self.message, UserMessage):
                raise HookContractError("UserPromptSubmit 需要 user message")
            if self.prepared is not None or self.result is not None or self.history or self.stop_hook_active:
                raise HookContractError("UserPromptSubmit 收到了其他事件的字段")
        elif self.event == "PreToolUse":
            if not _is_prepared(self.prepared):
                raise HookContractError("PreToolUse 需要有效的 prepared tool call")
            if self.message is not None or self.result is not None or self.history or self.stop_hook_active:
                raise HookContractError("PreToolUse 收到了其他事件的字段")
        elif self.event == "PostToolUse":
            if not _is_prepared(self.prepared) or not isinstance(self.result, ToolResult):
                raise HookContractError("PostToolUse 需要 prepared tool call 和 tool result")
            if self.message is not None or self.history or self.stop_hook_active:
                raise HookContractError("PostToolUse 收到了其他事件的字段")
        elif self.message is not None or self.prepared is not None or self.result is not None:
            raise HookContractError("Stop 收到了其他事件的字段")


@dataclass(frozen=True, slots=True)
class HookResult:
    """回调对循环提出的结构化影响。

    `additional_context` 只能是 system 消息，`force_continue` 只能是 user 消息；
    这样 Hook 无法伪造 assistant/tool 配对。所有返回值都复制成新对象，
    对应 Java 中从外部 DTO 转换成内部不可变值对象。
    """

    permission_behavior: PermissionBehavior = "passthrough"
    updated_input: PreparedToolCall | None = None
    updated_output: ToolResult | None = None
    additional_context: tuple[ChatMessage, ...] = ()
    blocking_error: ToolResult | None = None
    prevent_continuation: bool = False
    force_continue: UserMessage | None = None

    def __post_init__(self) -> None:
        if self.permission_behavior not in PERMISSION_BEHAVIORS:
            raise HookContractError("permission_behavior 必须是受支持的权限行为")
        if self.updated_input is not None and not _is_prepared(self.updated_input):
            raise HookContractError("updated_input 必须是有效的 prepared tool call")
        if self.updated_output is not None and not isinstance(self.updated_output, ToolResult):
            raise HookContractError("updated_output 必须是 ToolResult")
        if self.blocking_error is not None and (not isinstance(self.blocking_error, ToolResult) or not self.blocking_error.is_error):
            raise HookContractError("blocking_error 必须是错误 ToolResult")
        if not isinstance(self.additional_context, tuple) or not all(isinstance(item, SystemMessage) for item in self.additional_context):
            raise HookContractError("additional_context 只能包含 system 消息")
        if not isinstance(self.prevent_continuation, bool):
            raise HookContractError("prevent_continuation 必须是 bool")
        if self.force_continue is not None and not isinstance(self.force_continue, UserMessage):
            raise HookContractError("force_continue 必须是 user 消息")
        object.__setattr__(self, "updated_input", None if self.updated_input is None else copy_prepared_tool_call(self.updated_input))
        object.__setattr__(self, "updated_output", None if self.updated_output is None else copy_tool_result(self.updated_output))
        object.__setattr__(self, "additional_context", tuple(system_message(item.content or "") for item in self.additional_context))
        object.__setattr__(self, "blocking_error", None if self.blocking_error is None else copy_tool_result(self.blocking_error))
        object.__setattr__(self, "force_continue", None if self.force_continue is None else user_message(self.force_continue.content))

    def validate_for(self, event: HookEvent) -> None:
        """拒绝回调在错误事件上使用字段，类似 Java Bean Validation 的分组校验。"""
        invalid: list[str] = []
        if event != "PreToolUse":
            if self.permission_behavior != "passthrough": invalid.append("permission_behavior")
            if self.updated_input is not None: invalid.append("updated_input")
            if self.blocking_error is not None: invalid.append("blocking_error")
        if event != "PostToolUse":
            if self.updated_output is not None: invalid.append("updated_output")
            if self.prevent_continuation: invalid.append("prevent_continuation")
        if event != "Stop" and self.force_continue is not None:
            invalid.append("force_continue")
        if invalid:
            raise HookContractError(f"{event} HookResult 不允许字段: {', '.join(invalid)}")


HookCallback = Callable[[HookContext], HookResult | Awaitable[HookResult]]


class HookRegistry:
    """按注册顺序串行执行四类生命周期回调。"""

    def __init__(self) -> None:
        self._callbacks: dict[HookEvent, list[HookCallback]] = {event: [] for event in HOOK_EVENTS}

    def register(self, event: HookEvent, callback: HookCallback) -> None:
        """把回调追加到事件队列尾部；注册顺序就是执行顺序。"""
        if not _is_event(event):
            raise HookContractError("event 必须是受支持的 HookEvent")
        if not callable(callback):
            raise HookContractError("hook callback 必须可调用")
        self._callbacks[event].append(callback)

    async def run(self, context: HookContext) -> HookResult:
        """串行执行回调并把上一个回调的改写传给下一个回调。"""
        if not isinstance(context, HookContext):
            raise HookContractError("context 必须是 HookContext")
        combined = HookResult()
        current = context
        for callback in self._callbacks[context.event]:
            outcome = callback(current)
            if inspect.isawaitable(outcome):
                outcome = await outcome
            if not isinstance(outcome, HookResult):
                raise HookContractError(f"{context.event} hook callback 必须返回 HookResult")
            outcome.validate_for(context.event)
            normalized = self._normalize_input(current, outcome)
            if normalized is not None:
                outcome = HookResult(permission_behavior=outcome.permission_behavior, updated_input=normalized, additional_context=outcome.additional_context, blocking_error=outcome.blocking_error)
            if context.event == "Stop" and context.stop_hook_active and outcome.force_continue is not None:
                outcome = HookResult(additional_context=outcome.additional_context)
            combined = _merge_results(combined, outcome)
            if outcome.updated_input is not None:
                current = HookContext("PreToolUse", prepared=outcome.updated_input)
            elif outcome.updated_output is not None and current.prepared is not None:
                current = HookContext("PostToolUse", prepared=current.prepared, result=outcome.updated_output)
            if outcome.blocking_error is not None or outcome.force_continue is not None:
                break
        return combined

    async def run_user_prompt(self, message: UserMessage) -> HookResult:
        return await self.run(HookContext("UserPromptSubmit", message=message))

    async def run_pre_tool(self, prepared: PreparedToolCall) -> HookResult:
        return await self.run(HookContext("PreToolUse", prepared=prepared))

    async def run_post_tool(self, prepared: PreparedToolCall, result: ToolResult) -> HookResult:
        return await self.run(HookContext("PostToolUse", prepared=prepared, result=result))

    async def run_stop(self, history: Sequence[ChatMessage], stop_hook_active: bool) -> HookResult:
        return await self.run(HookContext("Stop", history=tuple(history), stop_hook_active=stop_hook_active))

    @staticmethod
    def _normalize_input(context: HookContext, result: HookResult) -> PreparedToolCall | None:
        updated = result.updated_input
        if updated is None:
            return None
        original = context.prepared
        if original is None or original.definition is None:
            raise HookContractError("updated_input 需要原 prepared tool call")
        if updated.call.id != original.call.id:
            raise HookContractError("updated_input 必须保留 tool call id")
        if updated.call.name != original.call.name:
            raise HookContractError("updated_input 必须保留工具名称")
        if updated.definition is not original.definition:
            raise HookContractError("updated_input 必须保留注册表中的工具定义")
        if original.definition.validator is not None and (updated.arguments is None or not original.definition.validator(updated.arguments)):
            raise HookContractError("updated_input 参数没有通过原工具 schema")
        return copy_prepared_tool_call(updated, definition=original.definition)


def _merge_results(current: HookResult, incoming: HookResult) -> HookResult:
    """合并多个回调：上下文累积、改写以后者为准、权限取最严格。"""
    return HookResult(
        permission_behavior=_stronger_permission(current.permission_behavior, incoming.permission_behavior),
        updated_input=incoming.updated_input or current.updated_input,
        updated_output=incoming.updated_output or current.updated_output,
        additional_context=current.additional_context + incoming.additional_context,
        blocking_error=current.blocking_error or incoming.blocking_error,
        prevent_continuation=current.prevent_continuation or incoming.prevent_continuation,
        force_continue=current.force_continue or incoming.force_continue,
    )


def _stronger_permission(current: PermissionBehavior, incoming: PermissionBehavior) -> PermissionBehavior:
    priority: Mapping[PermissionBehavior, int] = {"passthrough": 0, "allow": 1, "ask": 2, "deny": 3}
    return incoming if priority[incoming] > priority[current] else current

File: typescript/chapters/ch05/src/core/hooks.ts
import type { ChatMessage } from "./messages.js";
import { isChatMessage, systemMessage, userMessage } from "./messages.js";
import type { PermissionBehavior } from "./permissions.js";
import { isPermissionBehavior } from "./permissions.js";
import type { PreparedToolCall, ToolResult } from "./tools.js";
import { copyToolResult, freezePreparedToolCall, isToolResult } from "./tools.js";

export const HOOK_EVENTS = Object.freeze([
  "UserPromptSubmit",
  "PreToolUse",
  "PostToolUse",
  "Stop",
] as const);
// Hook 是受限扩展点：事件上下文和返回值均按事件类型严格校验。
export type HookEvent = (typeof HOOK_EVENTS)[number];

export class HookContractError extends Error {
  override readonly name: string = "HookContractError";
}

function isHookEvent(value: unknown): value is HookEvent {
  return HOOK_EVENTS.some((event) => event === value);
}

function isValidPrepared(prepared: unknown): prepared is PreparedToolCall {
  if (typeof prepared !== "object" || prepared === null) {
    return false;
  }
  const value = prepared as PreparedToolCall;
  const call = value.call;
  const definition = value.definition;
  return (
    typeof call === "object" &&
    call !== null &&
    typeof call.id === "string" &&
    call.id.length > 0 &&
    typeof call.name === "string" &&
    call.name.length > 0 &&
    typeof call.arguments === "string" &&
    value.error === undefined &&
    typeof definition === "object" &&
    definition !== null &&
    typeof definition.name === "string" &&
    typeof definition.inputSchema?.safeParse === "function" &&
    typeof definition.invoke === "function" &&
    value.arguments !== undefined
  );
}

export interface HookContextOptions {
  readonly event: HookEvent;
  readonly message?: ChatMessage;
  readonly prepared?: PreparedToolCall;
  readonly result?: ToolResult;
  readonly history?: readonly ChatMessage[];
  readonly stopHookActive?: boolean;
}

export class HookContext {
  // 每个事件只携带其拥有的数据，避免 Hook 误用其他生命周期阶段的状态。
  readonly event: HookEvent;
  readonly message: ChatMessage | undefined;
  readonly prepared: PreparedToolCall | undefined;
  readonly result: ToolResult | undefined;
  readonly history: readonly ChatMessage[];
  readonly stopHookActive: boolean;

  constructor(options: HookContextOptions) {
    if (!isHookEvent(options.event)) {
      throw new HookContractError("event must be a HookEvent");
    }
    const history = options.history === undefined ? [] : options.history;
    const stopHookActive = options.stopHookActive === undefined ? false : options.stopHookActive;
    if (!Array.isArray(history) || !history.every((message: unknown) => isChatMessage(message))) {
      throw new HookContractError(`${options.event} history must contain ChatMessage values`);
    }
    if (typeof stopHookActive !== "boolean") {
      throw new HookContractError("stopHookActive must be boolean");
    }

    this.event = options.event;
    this.message = options.message;
    this.prepared = options.prepared;
    this.result = options.result;
    this.history = Object.freeze([...history]);
    this.stopHookActive = stopHookActive;
    this.#validateEventFields();
    Object.freeze(this);
  }

  #validateEventFields(): void {
    if (this.event === "UserPromptSubmit") {
      if (!isChatMessage(this.message) || this.message.role !== "user") {
        throw new HookContractError("UserPromptSubmit requires a user message");
      }
      if (
        this.prepared !== undefined ||
        this.result !== undefined ||
        this.history.length > 0 ||
        this.stopHookActive
      ) {
        throw new HookContractError("UserPromptSubmit received fields owned by another event");
      }
      return;
    }
    if (this.event === "PreToolUse") {
      if (!isValidPrepared(this.prepared)) {
        throw new HookContractError("PreToolUse requires a valid prepared tool call");
      }
      if (
        this.message !== undefined ||
        this.result !== undefined ||
        this.history.length > 0 ||
        this.stopHookActive
      ) {
        throw new HookContractError("PreToolUse received fields owned by another event");
      }
      return;
    }
    if (this.event === "PostToolUse") {
      if (!isValidPrepared(this.prepared)) {
        throw new HookContractError("PostToolUse requires a valid prepared tool call");
      }
      if (!isToolResult(this.result)) {
        throw new HookContractError("PostToolUse requires a tool result");
      }
      if (this.message !== undefined || this.history.length > 0 || this.stopHookActive) {
        throw new HookContractError("PostToolUse received fields owned by another event");
      }
      return;
    }
    if (this.message !== undefined || this.prepared !== undefined || this.result !== undefined) {
      throw new HookContractError("Stop received fields owned by another event");
    }
  }
}

export interface HookResultOptions {
  readonly permissionBehavior?: PermissionBehavior;
  readonly updatedInput?: PreparedToolCall;
  readonly updatedOutput?: ToolResult;
  readonly additionalContext?: readonly ChatMessage[];
  readonly blockingError?: ToolResult;
  readonly preventContinuation?: boolean;
  readonly forceContinue?: ChatMessage;
}

export class HookResult {
  // 构造器内校验类型并冻结副本，防止回调引用扩散污染。
  // Hook 的副作用被建模为结构化结果，不能直接修改 Agent 内部状态。
  readonly permissionBehavior: PermissionBehavior;
  readonly updatedInput: PreparedToolCall | undefined;
  readonly updatedOutput: ToolResult | undefined;
  readonly additionalContext: readonly ChatMessage[];
  readonly blockingError: ToolResult | undefined;
  readonly preventContinuation: boolean;
  readonly forceContinue: ChatMessage | undefined;

  constructor(options: HookResultOptions = {}) {
    const permissionBehavior = options.permissionBehavior ?? "passthrough";
    const additionalContext = options.additionalContext ?? [];
    const preventContinuation = options.preventContinuation ?? false;
    if (!isPermissionBehavior(permissionBehavior)) {
      throw new HookContractError("permissionBehavior must be a PermissionBehavior");
    }
    if (options.updatedInput !== undefined && !isValidPrepared(options.updatedInput)) {
      throw new HookContractError("updatedInput must be a valid prepared tool call");
    }
    if (options.updatedOutput !== undefined && !isToolResult(options.updatedOutput)) {
      throw new HookContractError("updatedOutput must be a ToolResult");
    }
    if (
      options.blockingError !== undefined &&
      (!isToolResult(options.blockingError) || !options.blockingError.isError)
    ) {
      throw new HookContractError("blockingError must be an error ToolResult");
    }
    if (
      !Array.isArray(additionalContext) ||
      !additionalContext.every(
        (message: unknown) => isChatMessage(message) && message.role === "system",
      )
    ) {
      throw new HookContractError("additionalContext must contain system ChatMessage values");
    }
    if (typeof preventContinuation !== "boolean") {
      throw new HookContractError("preventContinuation must be boolean");
    }
    if (
      options.forceContinue !== undefined &&
      (!isChatMessage(options.forceContinue) || options.forceContinue.role !== "user")
    ) {
      throw new HookContractError("forceContinue must be a user message");
    }

    this.permissionBehavior = permissionBehavior;
    this.updatedInput = options.updatedInput;
    this.updatedOutput =
      options.updatedOutput === undefined ? undefined : copyToolResult(options.updatedOutput);
    this.additionalContext = Object.freeze(
      additionalContext.map((message) => systemMessage(message.content)),
    );
    this.blockingError =
      options.blockingError === undefined ? undefined : copyToolResult(options.blockingError);
    this.preventContinuation = preventContinuation;
    this.forceContinue =
      options.forceContinue === undefined ? undefined : userMessage(options.forceContinue.content);
    Object.freeze(this);
  }

  validateFor(event: HookEvent): void {
    if (!isHookEvent(event)) {
      throw new HookContractError("event must be a HookEvent");
    }
    const invalid: string[] = [];
    if (event !== "PreToolUse") {
      if (this.permissionBehavior !== "passthrough") {
        invalid.push("permissionBehavior");
      }
      if (this.updatedInput !== undefined) {
        invalid.push("updatedInput");
      }
      if (this.blockingError !== undefined) {
        invalid.push("blockingError");
      }
    }
    if (event !== "PostToolUse") {
      if (this.updatedOutput !== undefined) {
        invalid.push("updatedOutput");
      }
      if (this.preventContinuation) {
        invalid.push("preventContinuation");
      }
    }
    if (event !== "Stop" && this.forceContinue !== undefined) {
      invalid.push("forceContinue");
    }
    if (invalid.length > 0) {
      throw new HookContractError(
        `${event} HookResult does not allow fields: ${invalid.join(", ")}`,
      );
    }
  }
}

export type HookCallback = (context: HookContext) => HookResult | Promise<HookResult>;

export class HookRegistry {
  // 每个事件有独立回调队列，注册顺序即为执行顺序。
  // 回调按注册顺序合并；后续回调读取前一个回调规范化后的上下文。
  readonly #callbacks: Map<HookEvent, HookCallback[]> = new Map(
    HOOK_EVENTS.map((event) => [event, []]),
  );

  register(event: HookEvent, callback: HookCallback): void {
    if (!isHookEvent(event)) {
      throw new HookContractError("event must be a HookEvent");
    }
    if (typeof callback !== "function") {
      throw new HookContractError("hook callback must be callable");
    }
    const callbacks = this.#callbacks.get(event);
    if (callbacks === undefined) {
      throw new HookContractError(`hook registry is missing event: ${event}`);
    }
    callbacks.push(callback);
  }

  async run(context: HookContext): Promise<HookResult> {
    // 串行执行回调，合并结果；blockingError 或 forceContinue 短路。
    if (!(context instanceof HookContext)) {
      throw new HookContractError("context must be a HookContext");
    }
    let combined = new HookResult();
    let current = context;
    const callbacks = this.#callbacks.get(context.event);
    if (callbacks === undefined) {
      throw new HookContractError(`hook registry is missing event: ${context.event}`);
    }
    for (const callback of callbacks) {
      const outcome: unknown = await callback(current);
      if (!(outcome instanceof HookResult)) {
        throw new HookContractError(`${context.event} hook callback must return HookResult`);
      }
      outcome.validateFor(context.event);
      const normalizedInput = normalizeUpdatedInput(current, outcome);
      const normalizedOutcome =
        normalizedInput === undefined
          ? outcome
          : new HookResult({
              permissionBehavior: outcome.permissionBehavior,
              updatedInput: normalizedInput,
              additionalContext: outcome.additionalContext,
              ...(outcome.blockingError === undefined
                ? {}
                : { blockingError: outcome.blockingError }),
            });
      const effective =
        context.event === "Stop" &&
        context.stopHookActive &&
        normalizedOutcome.forceContinue !== undefined
          ? new HookResult({ additionalContext: normalizedOutcome.additionalContext })
          : normalizedOutcome;
      combined = mergeResults(combined, effective);

      if (effective.updatedInput !== undefined) {
        current = new HookContext({ event: "PreToolUse", prepared: effective.updatedInput });
      }
      if (effective.updatedOutput !== undefined && current.prepared !== undefined) {
        current = new HookContext({
          event: "PostToolUse",
          prepared: current.prepared,
          result: effective.updatedOutput,
        });
      }
      if (effective.blockingError !== undefined || effective.forceContinue !== undefined) {
        break;
      }
    }
    return combined;
  }

  async runUserPrompt(message: ChatMessage): Promise<HookResult> {
    return this.run(new HookContext({ event: "UserPromptSubmit", message }));
  }

  async runPreTool(prepared: PreparedToolCall): Promise<HookResult> {
    return this.run(new HookContext({ event: "PreToolUse", prepared }));
  }

  async runPostTool(prepared: PreparedToolCall, result: ToolResult): Promise<HookResult> {
    return this.run(new HookContext({ event: "PostToolUse", prepared, result }));
  }

  async runStop(history: readonly ChatMessage[], stopHookActive: boolean): Promise<HookResult> {
    return this.run(new HookContext({ event: "Stop", history, stopHookActive }));
  }
}

function normalizeUpdatedInput(
  // 重新解析 updatedInput 并冻结副本，防止“批准 A、执行 B”。
  context: HookContext,
  result: HookResult,
): PreparedToolCall | undefined {
  const updated = result.updatedInput;
  if (updated === undefined) {
    return undefined;
  }
  const original = context.prepared;
  if (original === undefined || original.definition === undefined) {
    throw new HookContractError("updatedInput requires an existing prepared tool call");
  }
  if (updated.call.id !== original.call.id) {
    throw new HookContractError("updatedInput must preserve the OpenAI tool call id");
  }
  if (updated.call.name !== original.call.name) {
    throw new HookContractError("updatedInput must preserve the tool name");
  }
  if (updated.definition !== original.definition) {
    throw new HookContractError("updatedInput must preserve the registered definition");
  }
  const parsed = original.definition.inputSchema.safeParse(updated.arguments);
  if (!parsed.success) {
    throw new HookContractError("updatedInput arguments must match the registered input schema");
  }
  // 审批和执行只读取这个脱离 Hook 原引用的冻结副本，避免批准 A、执行 B。
  return freezePreparedToolCall(updated.call, original.definition, parsed.data);
}

function mergeResults(current: HookResult, incoming: HookResult): HookResult {
  // 合并策略：updatedInput/Output 以后优先，additionalContext 串联，
  // preventContinuation OR，permissionBehavior 取最严格，
  // blockingError/forceContinue 短路并保留最先出现的。
  return new HookResult({
    permissionBehavior: strongerPermission(current.permissionBehavior, incoming.permissionBehavior),
    ...(incoming.updatedInput === undefined
      ? current.updatedInput === undefined
        ? {}
        : { updatedInput: current.updatedInput }
      : { updatedInput: incoming.updatedInput }),
    ...(incoming.updatedOutput === undefined
      ? current.updatedOutput === undefined
        ? {}
        : { updatedOutput: current.updatedOutput }
      : { updatedOutput: incoming.updatedOutput }),
    additionalContext: [...current.additionalContext, ...incoming.additionalContext],
    ...(incoming.blockingError === undefined
      ? current.blockingError === undefined
        ? {}
        : { blockingError: current.blockingError }
      : { blockingError: incoming.blockingError }),
    preventContinuation: current.preventContinuation || incoming.preventContinuation,
    ...(incoming.forceContinue === undefined
      ? current.forceContinue === undefined
        ? {}
        : { forceContinue: current.forceContinue }
      : { forceContinue: incoming.forceContinue }),
  });
}

function strongerPermission(
  current: PermissionBehavior,
  incoming: PermissionBehavior,
): PermissionBehavior {
  const priority: Readonly<Record<PermissionBehavior, number>> = {
    passthrough: 0,
    allow: 1,
    ask: 2,
    deny: 3,
  };
  return priority[incoming] > priority[current] ? incoming : current;
}
